### ISSUE TITLE
Pickle error

### ISSUE THREAD

**Issue by steve-solun** [Delgan/loguru#753]
Hey team,

Getting the following error, how can I solve it? Parsing issues?

```
--- Logging error in Loguru Handler #1 ---
Record was: {'elapsed': datetime.timedelta(seconds=663, microseconds=27414), 'exception': None, 'extra': {'output': [<Record n=<Node element_id='7143' labels=frozenset({'ECSCluster'}) properties={'firstseen': 1662903107411, 'name': 'synchronograph', 'lastupdated': 1662905395, 'settings_container_insights': 'disabled', 'id': 'arn:aws:ecs:eu-west-1:564407622114:cluster/synchronograph', 'region': 'eu-west-1', 'arn': 'arn:aws:ecs:eu-west-1:564407622114:cluster/synchronograph', 'capacity_providers': [], 'status': 'ACTIVE'}> r=<Relationship element_id='7849' nodes=(<Node element_id='7143' labels=frozenset({'ECSCluster'}) properties={'firstseen': 1662903107411, 'name': 'synchronograph', 'lastupdated': 1662905395, 'settings_container_insights': 'disabled', 'id': 'arn:aws:ecs:eu-west-1:564407622114:cluster/synchronograph', 'region': 'eu-west-1', 'arn': 'arn:aws:ecs:eu-west-1:564407622114:cluster/synchronograph', 'capacity_providers': [], 'status': 'ACTIVE'}>, <Node element_id='7222' labels=frozenset({'ECSService'}) properties={'enable_execute_command': False, 'cluster_arn': 'arn:aws:ecs:eu-west-1:564407622114:cluster/synchronograph', 'enable_ecs_managed_tags': False, 'created_at': 1647794867, 'desired_count': 0, 'created_by': 'arn:aws:iam::564407622114:user/costa@kubiya.io', 'propagate_tags': 'NONE', 'deployment_config_minimum_healthy_percent': 100, 'launch_type': 'EC2', 'firstseen': 1662903120686, 'pending_count': 0, 'deployment_config_circuit_breaker_enable': False, 'task_definition': 'arn:aws:ecs:eu-west-1:564407622114:task-definition/Nginx:1', 'name': 'synchronograph-service', 'lastupdated': 1662905395, 'deployment_config_circuit_breaker_rollback': False, 'running_count': 0, 'id': 'arn:aws:ecs:eu-west-1:564407622114:service/synchronograph/synchronograph-service', 'region': 'eu-west-1', 'arn': 'arn:aws:ecs:eu-west-1:564407622114:service/synchronograph/synchronograph-service', 'deployment_config_maximum_percent': 200, 'status': 'ACTIVE'}>) type='HAS_SERVICE' properties={'firstseen': 1662903120686, 'lastupdated': 1662905395}> k=<Node element_id='7222' labels=frozenset({'ECSService'}) properties={'enable_execute_command': False, 'cluster_arn': 'arn:aws:ecs:eu-west-1:564407622114:cluster/synchronograph', 'enable_ecs_managed_tags': False, 'created_at': 1647794867, 'desired_count': 0, 'created_by': 'arn:aws:iam::564407622114:user/costa@kubiya.io', 'propagate_tags': 'NONE', 'deployment_config_minimum_healthy_percent': 100, 'launch_type': 'EC2', 'firstseen': 1662903120686, 'pending_count': 0, 'deployment_config_circuit_breaker_enable': False, 'task_definition': 'arn:aws:ecs:eu-west-1:564407622114:task-definition/Nginx:1', 'name': 'synchronograph-service', 'lastupdated': 1662905395, 'deployment_config_circuit_breaker_rollback': False, 'running_count': 0, 'id': 'arn:aws:ecs:eu-west-1:564407622114:service/synchronograph/synchronograph-service', 'region': 'eu-west-1', 'arn': 'arn:aws:ecs:eu-west-1:564407622114:service/synchronograph/synchronograph-service', 'deployment_config_maximum_percent': 200, 'status': 'ACTIVE'}> n_attr_scores=[{'score': 0.8333, 'node_labels': ['ECSCluster'], 'value': 'synchronograph', 'key': 'name', 'node_id': 7143}]>, <Record n=<Node element_id='7143' labels=frozenset({'ECSCluster'}) properties={'firstseen': 1662903107411, 'name': 'synchronograph', 'lastupdated': 1662905395, 'settings_container_insights': 'disabled', 'id': 'arn:aws:ecs:eu-west-1:564407622114:cluster/synchronograph', 'region': 'eu-west-1', 'arn': 'arn:aws:ecs:eu-west-1:564407622114:cluster/synchronograph', 'capacity_providers': [], 'status': 'ACTIVE'}> r=<Relationship element_id='7847' nodes=(<Node element_id='7143' labels=frozenset({'ECSCluster'}) properties={'firstseen': 1662903107411, 'name': 'synchronograph', 'lastupdated': 1662905395, 'settings_container_insights': 'disabled', 'id': 'arn:aws:ecs:eu-west-1:564407622114:cluster/synchronograph', 'region': 'eu-west-1', 'arn': 'arn:aws:ecs:eu-west-1:564407622114:cluster/synchronograph', 'capacity_providers': [], 'status': 'ACTIVE'}>, <Node element_id='7221' labels=frozenset({'ECSService'}) properties={'enable_execute_command': False, 'cluster_arn': 'arn:aws:ecs:eu-west-1:564407622114:cluster/synchronograph', 'enable_ecs_managed_tags': False, 'created_at': 1647794868, 'desired_count': 0, 'created_by': 'arn:aws:iam::564407622114:user/costa@kubiya.io', 'propagate_tags': 'NONE', 'deployment_config_minimum_healthy_percent': 100, 'launch_type': 'EC2', 'firstseen': 1662903120686, 'pending_count': 0, 'deployment_config_circuit_breaker_enable': False, 'task_definition': 'arn:aws:ecs:eu-west-1:564407622114:task-definition/Nginx:1', 'name': 'synchronograph-service-2', 'lastupdated': 1662905395, 'deployment_config_circuit_breaker_rollback': False, 'running_count': 0, 'id': 'arn:aws:ecs:eu-west-1:564407622114:service/synchronograph/synchronograph-service-2', 'region': 'eu-west-1', 'arn': 'arn:aws:ecs:eu-west-1:564407622114:service/synchronograph/synchronograph-service-2', 'deployment_config_maximum_percent': 200, 'status': 'ACTIVE'}>) type='HAS_SERVICE' properties={'firstseen': 1662903120686, 'lastupdated': 1662905395}> k=<Node element_id='7221' labels=frozenset({'ECSService'}) properties={'enable_execute_command': False, 'cluster_arn': 'arn:aws:ecs:eu-west-1:564407622114:cluster/synchronograph', 'enable_ecs_managed_tags': False, 'created_at': 1647794868, 'desired_count': 0, 'created_by': 'arn:aws:iam::564407622114:user/costa@kubiya.io', 'propagate_tags': 'NONE', 'deployment_config_minimum_healthy_percent': 100, 'launch_type': 'EC2', 'firstseen': 1662903120686, 'pending_count': 0, 'deployment_config_circuit_breaker_enable': False, 'task_definition': 'arn:aws:ecs:eu-west-1:564407622114:task-definition/Nginx:1', 'name': 'synchronograph-service-2', 'lastupdated': 1662905395, 'deployment_config_circuit_breaker_rollback': False, 'running_count': 0, 'id': 'arn:aws:ecs:eu-west-1:564407622114:service/synchronograph/synchronograph-service-2', 'region': 'eu-west-1', 'arn': 'arn:aws:ecs:eu-west-1:564407622114:service/synchronograph/synchronograph-service-2', 'deployment_config_maximum_percent': 200, 'status': 'ACTIVE'}> n_attr_scores=[{'score': 0.8333, 'node_labels': ['ECSCluster'], 'value': 'synchronograph', 'key': 'name', 'node_id': 7143}]>]}, 'file': (name='utils.py', path='/Users/stevesolun/Steves_Files/Code/resource_search/src/utils.py'), 'function': 'query_engine', 'level': (name='INFO', no=20, icon='ℹ️'), 'line': 145, 'message': "Exiting `query_engine` function with result = [<Record n=<Node element_id='7143' labels=frozenset({'ECSCluster'}) properties={'firstseen': 1662903107411, 'name': 'synchronograph', 'lastupdated': 1662905395, 'settings_container_insights': 'disabled', 'id': 'arn:aws:ecs:eu-west-1:564407622114:cluster/synchronograph', 'region': 'eu-west-1', 'arn': 'arn:aws:ecs:eu-west-1:564407622114:cluster/synchronograph', 'capacity_providers': [], 'status': 'ACTIVE'}> r=<Relationship element_id='7849' nodes=(<Node element_id='7143' labels=frozenset({'ECSCluster'}) properties={'firstseen': 1662903107411, 'name': 'synchronograph', 'lastupdated': 1662905395, 'settings_container_insights': 'disabled', 'id': 'arn:aws:ecs:eu-west-1:564407622114:cluster/synchronograph', 'region': 'eu-west-1', 'arn': 'arn:aws:ecs:eu-west-1:564407622114:cluster/synchronograph', 'capacity_providers': [], 'status': 'ACTIVE'}>, <Node element_id='7222' labels=frozenset({'ECSService'}) properties={'enable_execute_command': False, 'cluster_arn': 'arn:aws:ecs:eu-west-1:564407622114:cluster/synchronograph', 'enable_ecs_managed_tags': False, 'created_at': 1647794867, 'desired_count': 0, 'created_by': 'arn:aws:iam::564407622114:user/costa@kubiya.io', 'propagate_tags': 'NONE', 'deployment_config_minimum_healthy_percent': 100, 'launch_type': 'EC2', 'firstseen': 1662903120686, 'pending_count': 0, 'deployment_config_circuit_breaker_enable': False, 'task_definition': 'arn:aws:ecs:eu-west-1:564407622114:task-definition/Nginx:1', 'name': 'synchronograph-service', 'lastupdated': 1662905395, 'deployment_config_circuit_breaker_rollback': False, 'running_count': 0, 'id': 'arn:aws:ecs:eu-west-1:564407622114:service/synchronograph/synchronograph-service', 'region': 'eu-west-1', 'arn': 'arn:aws:ecs:eu-west-1:564407622114:service/synchronograph/synchronograph-service', 'deployment_config_maximum_percent': 200, 'status': 'ACTIVE'}>) type='HAS_SERVICE' properties={'firstseen': 1662903120686, 'lastupdated': 1662905395}> k=<Node element_id='7222' labels=frozenset({'ECSService'}) properties={'enable_execute_command': False, 'cluster_arn': 'arn:aws:ecs:eu-west-1:564407622114:cluster/synchronograph', 'enable_ecs_managed_tags': False, 'created_at': 1647794867, 'desired_count': 0, 'created_by': 'arn:aws:iam::564407622114:user/costa@kubiya.io', 'propagate_tags': 'NONE', 'deployment_config_minimum_healthy_percent': 100, 'launch_type': 'EC2', 'firstseen': 1662903120686, 'pending_count': 0, 'deployment_config_circuit_breaker_enable': False, 'task_definition': 'arn:aws:ecs:eu-west-1:564407622114:task-definition/Nginx:1', 'name': 'synchronograph-service', 'lastupdated': 1662905395, 'deployment_config_circuit_breaker_rollback': False, 'running_count': 0, 'id': 'arn:aws:ecs:eu-west-1:564407622114:service/synchronograph/synchronograph-service', 'region': 'eu-west-1', 'arn': 'arn:aws:ecs:eu-west-1:564407622114:service/synchronograph/synchronograph-service', 'deployment_config_maximum_percent': 200, 'status': 'ACTIVE'}> n_attr_scores=[{'score': 0.8333, 'node_labels': ['ECSCluster'], 'value': 'synchronograph', 'key': 'name', 'node_id': 7143}]>, <Record n=<Node element_id='7143' labels=frozenset({'ECSCluster'}) properties={'firstseen': 1662903107411, 'name': 'synchronograph', 'lastupdated': 1662905395, 'settings_container_insights': 'disabled', 'id': 'arn:aws:ecs:eu-west-1:564407622114:cluster/synchronograph', 'region': 'eu-west-1', 'arn': 'arn:aws:ecs:eu-west-1:564407622114:cluster/synchronograph', 'capacity_providers': [], 'status': 'ACTIVE'}> r=<Relationship element_id='7847' nodes=(<Node element_id='7143' labels=frozenset({'ECSCluster'}) properties={'firstseen': 1662903107411, 'name': 'synchronograph', 'lastupdated': 1662905395, 'settings_container_insights': 'disabled', 'id': 'arn:aws:ecs:eu-west-1:564407622114:cluster/synchronograph', 'region': 'eu-west-1', 'arn': 'arn:aws:ecs:eu-west-1:564407622114:cluster/synchronograph', 'capacity_providers': [], 'status': 'ACTIVE'}>, <Node element_id='7221' labels=frozenset({'ECSService'}) properties={'enable_execute_command': False, 'cluster_arn': 'arn:aws:ecs:eu-west-1:564407622114:cluster/synchronograph', 'enable_ecs_managed_tags': False, 'created_at': 1647794868, 'desired_count': 0, 'created_by': 'arn:aws:iam::564407622114:user/costa@kubiya.io', 'propagate_tags': 'NONE', 'deployment_config_minimum_healthy_percent': 100, 'launch_type': 'EC2', 'firstseen': 1662903120686, 'pending_count': 0, 'deployment_config_circuit_breaker_enable': False, 'task_definition': 'arn:aws:ecs:eu-west-1:564407622114:task-definition/Nginx:1', 'name': 'synchronograph-service-2', 'lastupdated': 1662905395, 'deployment_config_circuit_breaker_rollback': False, 'running_count': 0, 'id': 'arn:aws:ecs:eu-west-1:564407622114:service/synchronograph/synchronograph-service-2', 'region': 'eu-west-1', 'arn': 'arn:aws:ecs:eu-west-1:564407622114:service/synchronograph/synchronograph-service-2', 'deployment_config_maximum_percent': 200, 'status': 'ACTIVE'}>) type='HAS_SERVICE' properties={'firstseen': 1662903120686, 'lastupdated': 1662905395}> k=<Node element_id='7221' labels=frozenset({'ECSService'}) properties={'enable_execute_command': False, 'cluster_arn': 'arn:aws:ecs:eu-west-1:564407622114:cluster/synchronograph', 'enable_ecs_managed_tags': False, 'created_at': 1647794868, 'desired_count': 0, 'created_by': 'arn:aws:iam::564407622114:user/costa@kubiya.io', 'propagate_tags': 'NONE', 'deployment_config_minimum_healthy_percent': 100, 'launch_type': 'EC2', 'firstseen': 1662903120686, 'pending_count': 0, 'deployment_config_circuit_breaker_enable': False, 'task_definition': 'arn:aws:ecs:eu-west-1:564407622114:task-definition/Nginx:1', 'name': 'synchronograph-service-2', 'lastupdated': 1662905395, 'deployment_config_circuit_breaker_rollback': False, 'running_count': 0, 'id': 'arn:aws:ecs:eu-west-1:564407622114:service/synchronograph/synchronograph-service-2', 'region': 'eu-west-1', 'arn': 'arn:aws:ecs:eu-west-1:564407622114:service/synchronograph/synchronograph-service-2', 'deployment_config_maximum_percent': 200, 'status': 'ACTIVE'}> n_attr_scores=[{'score': 0.8333, 'node_labels': ['ECSCluster'], 'value': 'synchronograph', 'key': 'name', 'node_id': 7143}]>]", 'module': 'utils', 'name': 'utils', 'process': (id=84591, name='MainProcess'), 'thread': (id=8499209472, name='MainThread'), 'time': datetime(2022, 11, 23, 11, 46, 21, 278357, tzinfo=datetime.timezone(datetime.timedelta(seconds=7200), 'IST'))}
Traceback (most recent call last):
  File "/Users/user/.local/share/virtualenvs/cccccccc-/lib/python3.8/site-packages/loguru/_handler.py", line 175, in emit
    self._queue.put(str_record)
  File "/opt/homebrew/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/multiprocessing/queues.py", line 362, in put
    obj = _ForkingPickler.dumps(obj)
  File "/opt/homebrew/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/multiprocessing/reduction.py", line 51, in dumps
    cls(buf, protocol).dump(obj)
_pickle.PicklingError: Can't pickle <class 'abc.HAS_SERVICE'>: attribute lookup HAS_SERVICE on abc failed


```

**Comment by Delgan** [comment]
Hi.

Your logged message contains an object which is not picklable, this is a problem because the message must be serialized using `pickle` when `enqueue=True`.

You have to find out where such an object is likely to be logged and remove it, or implement the pickling protocol correctly.

**Comment by Delgan** [comment]
This issue has been fixed by #905, therefore I'm closing it.
